src/main/java/io/vertx/ext/amqp/impl/AmqpReceiverImpl.java

```diff
@@ -0,0 +1,342 @@
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+package io.vertx.ext.amqp.impl;
+
+import io.vertx.codegen.annotations.Nullable;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.VertxException;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.amqp.AmqpConnection;
+import io.vertx.ext.amqp.AmqpMessage;
+import io.vertx.ext.amqp.AmqpReceiver;
+import io.vertx.ext.amqp.AmqpReceiverOptions;
+import io.vertx.proton.ProtonReceiver;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+public class AmqpReceiverImpl implements AmqpReceiver {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AmqpReceiverImpl.class);
+
+  /**
+   * The address.
+   * Not final because for dynamic link the address is set when the createReceiver is opened.
+   */
+  private String address;
+  private final ProtonReceiver receiver;
+  private final AmqpConnectionImpl connection;
+  private final Queue<AmqpMessageImpl> buffered = new ArrayDeque<>();
+
+  private Handler<AmqpMessage> handler;
+  private long demand = Long.MAX_VALUE;
+  private boolean closed;
+  private Handler<Void> endHandler;
+  private Handler<Throwable> exceptionHandler;
+  private boolean initialCreditGiven;
+  private int initialCredit = 1000;
+  private final boolean durable;
+
+  /**
+   * Creates a new instance of {@link AmqpReceiverImpl}.
+   * This method must be called on the connection context.
+   *
+   * @param address           the address, may be {@code null} for dynamic links
+   * @param connection        the connection
+   * @param options           the receiver options, must not be {@code null}
+   * @param receiver          the underlying proton createReceiver
+   * @param handler           the handler
+   * @param completionHandler called when the createReceiver is opened
+   */
+  AmqpReceiverImpl(
+    String address,
+    AmqpConnectionImpl connection,
+    AmqpReceiverOptions options,
+    ProtonReceiver receiver,
+    Handler<AmqpMessage> handler, Handler<AsyncResult<AmqpReceiver>> completionHandler) {
+    this.address = address;
+    this.receiver = receiver;
+    this.connection = connection;
+    this.handler = handler;
+    this.durable = options.isDurable();
+    int maxBufferedMessages = options.getMaxBufferedMessages();
+    if (maxBufferedMessages > 0) {
+      this.initialCredit = maxBufferedMessages;
+    }
+
+    // Disable auto-accept and automated prefetch, we manage disposition and credit
+    // manually to allow for delayed handler registration and pause/resume functionality.
+    this.receiver
+      .setAutoAccept(false)
+      .setPrefetch(0);
+
+    this.receiver.handler((delivery, message) -> handleMessage(new AmqpMessageImpl(message, delivery)));
+    if (this.handler != null) {
+      handler(this.handler);
+    }
+
+    this.receiver.closeHandler(res -> {
```

Comment:
This handles receipt of frame detach(closed=true), there is a related detachHandler for handling detach(closed=false)that should also be doing much the same thing here.

Need to vary whether to call receiver.close() or receiver.detach() as the AmqpReceiverImpl#close(..) method is doing.

Reply:
I've implemented the detachHandler.